source/base/DateTime.ooc

```diff
@@ -14,22 +14,246 @@
+
+	_ticksToDateTimeHelper: static func (totalTicks: Int64) -> DateTimeData {
+		result := DateTimeData new()
+		fourYearBlocks := totalTicks / This TicksPerFourYears
+		year := 4 * fourYearBlocks
+		ticksLeft := totalTicks - fourYearBlocks * This TicksPerFourYears
+		for (y in year + 1 .. year + 5) {
+			t := This ticksInYear(y)
+			if (ticksLeft < t) {
+				year = y
+				break
+			} else {
+				ticksLeft -= t
+			}
+		}
+		month := 0
+		for (m in 1 .. 13) {
+			t := This ticksInMonth(year, m)
+			if (ticksLeft < t) {
+				month = m
+				break
+			} else {
+				ticksLeft -= t
+			}
+		}
+		days := ticksLeft / This TicksPerDay
+		ticksLeft -= days * This TicksPerDay
+		hour := ticksLeft / This TicksPerHour
+		ticksLeft -= hour * This TicksPerHour
+		minute := ticksLeft / This TicksPerMinute
+		ticksLeft -= minute * This TicksPerMinute
+		second := ticksLeft / This TicksPerSecond
+		ticksLeft -= second * This TicksPerSecond
+		millisecond := ticksLeft / This TicksPerMillisecond
+		result year = year
+		result month = month
+		result day = days + 1
+		result hour = hour
+		result minute = minute
+		result second = second
+		result millisecond = millisecond
+		result
+	}
+
+	/* returns number of ticks for given hours, minutes and seconds */
+	timeToTicks: static func(hours, minutes, seconds, millisecond: Int) -> Int64 {
+		(hours * 3600 + minutes * 60 + seconds) * This TicksPerSecond + millisecond * This TicksPerMillisecond
+	}
+
+	/* returns number of ticks for given date at 0:00*/
+	dateToTicks: static func(year, month, day : Int) -> UInt64 {
+		if (This dateIsValid(year, month, day)) {
+			totalDays := day - 1
+			for (m in 1 .. month) {
+				totalDays += This daysInMonth(year, m)
+			}
+			fourYearBlocks := (year - 1) / 4
+			year_start := fourYearBlocks * 4
+			for (y in year_start + 1 .. year)
+				totalDays += This daysInYear(y)
+			totalDays += fourYearBlocks * This DaysInFourYears
+			totalDays * This TicksPerDay
+		} else {
+			0
+		}
+	}
+
+	daysInYear: static func (year: Int) -> Int {
+		This DaysInYear + isLeapYear(year)
+	}
+
+	ticksInYear: static func (year: Int) -> UInt64 {
+		This daysInYear(year) * This TicksPerDay
+	}
+
+	daysInMonth: static func (year, month: Int) -> Int {
+		if (month == 2)
+			isLeapYear(year) ? 29 : 28
+		else if (month <= 7)
+			month % 2 ? 31 : 30
+		else
+			month % 2 ? 30 : 31
+	}
+
+	ticksInMonth: static func (year, month: Int) -> UInt64 {
+		This TicksPerDay * This daysInMonth(year, month)
+	}
+	/* validate argument ranges for hour/minutes/seconds vaules */
+	timeIsValid: static func (hour, minute, second, millisecond: Int) -> Bool {
+		hour >= 0 && hour < 24 && minute >= 0 && minute < 60 && second >= 0 && second < 60 && millisecond >= 0 && millisecond < 1000
+	}
+	/* validate argument ranges for year/month/day values */
+	dateIsValid: static func (year, month, day : Int) -> Bool {
+		year >= 1 && month >= 1 && month <= 12 && day >= 1 && day <= daysInMonth(year, month)
```

Comment:
By the same rationale, the `DaysInFourYears` constant also shouldn't cause much trouble since it's true for the years 1900-2100...